presto-spi/src/main/java/com/facebook/presto/spi/block/SingleMapBlock.java

```diff
@@ -35,24 +34,14 @@
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(SingleMapBlock.class).instanceSize();
 
     private final int offset;
-    private final int positionCount;
-    private final Block keyBlock;
-    private final Block valueBlock;
-    private final int[] hashTable;
-    final Type keyType;
-    private final MethodHandle keyNativeHashCode;
-    private final MethodHandle keyBlockNativeEquals;
-
-    SingleMapBlock(int offset, int positionCount, Block keyBlock, Block valueBlock, int[] hashTable, Type keyType, MethodHandle keyNativeHashCode, MethodHandle keyBlockNativeEquals)
+    private final int positionCount;  // The number of keys in this single map * 2
+    private final AbstractMapBlock mapBlock;
+
+    SingleMapBlock(int offset, int positionCount, AbstractMapBlock mapBlock)
```

Comment:
If we were to introduce an interface for this class, what would the API look like for that interface?  If it is only a small number of methods, we might want to add something like that to keep the abstractions between these classes simpler.  I'm not saying we should actually add an interface here, yet; I'm just curious what it would look like if we did.

Reply:
@dain which classes are you considering to implement this interface? Did you just mean SingleMapBlock and AbstractSingleMapBlock?

Reply:
I mean the `AbstractSingleMapBlock` argument here.  If that were an interface build specifically for this class, what methods would it have?  If it is small, we might want to add one to clean up the code and simplify testing.... just a thought

Reply:
@dain Did you mean the mapBlock we passed in? It's AbstractMapBlock. The referenced methods in SingleMapBlock include getRawKeyBlock(), getRawValueBlock(), getHashTables(), and it also accesses keyNativeHashCode and keyBlockNativeEquals members. Do you think it's worth making a new interface? If yes I'll add getKeyNativeHashCode() and getKeyBlockNativeEquals() and make a 5 method interface in a separate commit.

Reply:
If it is only 5 methods, I would consider adding the interface.  This is just my opinion... I generally get a bad feeling anytime I see a method taking a parameter with a type named `Abstract*`, as to me it screams, we should have an interface here.  In this case I would name the interface `MapBlockData`.  Again, just my opinion.  You can ask others how they feel about this.

Reply:
@dain makes sense. Shall I send a new PR for this interface or a seperate commit in this same PR, or just in this commit?

Reply:
All of those options are fine with me.  Maybe ask @haozhun what he prefers.

Reply:
The problem here is that it cannot be an interface because those fields are not public.